raiden/network/proxies/README.md

```diff
@@ -1,14 +1,266 @@
+2. Gas estimation: If the preconditions are met, then the transaction's
+   `startgas` is estimated through `eth_estimateGas` against the `latest` block
+   [[9]](#9). If the gas estimation failed. A race condition is *assumed* and
+   the reason for the failue is checked. This checks if the transaction is
+   still valid against the **current** state of the smart contract, and not the
+   state at triggering block.
+3. Transaction is sent: If the gas estimation was successful, then the
+   transaction is sent. The transaction may fail in two ways:
+    1. With an assert: This either a smart contract (a check that should be a
+       require and not an assert), compiler (a bad instruction was emitted), or
+       estimate gas bug (the assert was hit but the estimation gas did not
+       fail).
+    2. With a revert: A race condition is assumed, and the side-effects are
+       checked. If the smart contract state did not change a bug is assumed.
+
+### Preconditions
+
+The preconditions checks can fail on some edge cases, namely when a node goes
+offline for an extended period of time. If user requested a deposit at block
+`n`, goes offline and only comes back after `prune_blocks` time, the
+precondition checks cannot be done.
+
+As a consequence from the above, it's not always possible to detect client
+errors on edge cases. The approach taken in the current version of the code is
+to run:
+
+```python
+try:
+    # First query all relevant data for the checks
+    first = proxy.get_first(at=triggering_block.block_number)
+    second = proxy.get_second(at=triggering_block.block_number)
+    nth = proxy.get_n(at=triggering_block.block_number)
+    block = self.get_block(block=triggerring_block.block_number)
+except ValueError:
+    # if any of the above fails, the checks cannot be performed.
+    # Note that the first call may succeed, but the second can
+    # still fail, because of race conditions.
+    pass
+else:
+    assert block['blockHash'] == trigerring_block.block_hash
+    # if all data is available perform the checks
+    precondition_check(first, second, nth)
+```
+
+It is also importanto to note that precondition checks are only done for one
+transaction at time. A precondition check by itself can not prevent nor detect
+race conditions with other nodes in the network or even with the node itself.
+[[12]](#12)
+
+### Gas estimation
+
+Gas estimation is assumed to fail under these circumstances [[10]](#10)
+[[11]](#11):
+
+- The state of the smart contract changed since the `triggering_block`, and a
+  requirement is not met anymore. IOW, a race condition against the
+  participants transactions.
+- The smart contract or its compiler had a bug. Not all preconditions were
+  covered: this is a client bug.
+
+The first hypothesis can be checked if the state machine for the smart contract
+is a DAG. The state of the smart contract will have to have moved forward on
+the `latest` block (see below).
+
+```graphviz
+digraph {
```

Comment:
Is that actually rendered somewhere?

Reply:
github doesn't seem to support it, it was a feature of hackmd